scripts/qcow_util.py

```diff
@@ -0,0 +1,536 @@
+        with open(filename, "rb") as qcow2_file:
+            self.filename = filename  # Keep the filename if clean is called
+            self.header = self._read_qcow2_header(qcow2_file)
+            self.l1 = self._get_l1_entries(qcow2_file)
+            # The l1_to_l2 allows to get L2 entries for a given L1. If L1 entry
+            # is not allocated we store an empty list.
+            self.l1_to_l2: Dict[int, List[int]] = {}
+
+            for l1_entry in self.l1:
+                l2_offset = l1_entry & QcowInfo.L2_OFFSET_MASK
+                if l2_offset == 0:
+                    self.l1_to_l2[l1_entry] = []
+                else:
+                    self.l1_to_l2[l1_entry] = self._get_l2_entries(
+                        qcow2_file, l2_offset
+                    )
+
+    @staticmethod
+    def _move_backing_file(
+        f: BinaryIO, old_offset: int, new_offset: int, data_size: int
+    ) -> None:
+        """Move a number of bytes from old_offset to new_offset and replaces the old
+           value by 0s. It is up to the caller to save the current position in the file
+           if needed.
+
+        Args:
+            f: the file the will be modified
+            old_offset: the current offset
+            new_offset: the new offset where we want to move data
+            data_size: Size in bytes of data that we want to move
+
+        Returns:
+            Nothing but the file f is modified and the position in the file also.
+        """
+        # Read the string at backing_file_offset
+        f.seek(old_offset)
+        data = f.read(data_size)
+
+        # Write zeros at the original location
+        f.seek(old_offset)
+        f.write(b"\x00" * data_size)
+
+        # Write the string to the new location
+        f.seek(new_offset)
+        f.write(data)
+
+    def _add_or_find_custom_header(self) -> int:
+        """Add custom header at the end of header extensions
+
+        It finds the end of the header extensions and add the custom header.
+        If the header already exists nothing is done.
+
+        Args:
+
+        Returns:
+            It returns the data offset where custom header is found or created.
+            If data offset is 0 something weird happens.
+            The qcow2 file in self.filename can be modified.
+        """
+        header_length = 72  # This is the default value for version 2 images
+
+        custom_header_type = 0x76617465  # vate: it is easy to recognize with hexdump -C
+        custom_header_length = 8
+        custom_header_data = 0
+        # We don't need padding because we are already aligned
+        custom_header = struct.pack(
+            ">IIQ", custom_header_type, custom_header_length, custom_header_data
+        )
+
+        with open(self.filename, "rb+") as qcow2_file:
+            if self.header["version"] == 3:
+                qcow2_file.seek(100)  # 100 is the offset of header_length
+                header_length = int.from_bytes(qcow2_file.read(4))
```

Comment:
It looks like we need to add the endianess: `header_length = int.from_bytes(qcow2_file.read(4), 'big')`